cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRunner.java

```diff
@@ -180,12 +209,59 @@ public ProgramController run(Program program, ProgramOptions options) {
     }
   }
 
+  @Override
+  public FilterClassLoader.Filter getFilter() {
+    return SPARK_PROGRAM_CLASS_LOADER_FILTER;
+  }
+
+  /**
+   * Creates a {@link Closeable} for closing the ClassLoader of this {@link SparkProgramRunner}. The
+   * ClassLoader needs to be closed because there is one such ClassLoader created per program execution by
+   * the {@link SparkProgramRuntimeProvider} to support concurrent Spark program execution in the same JVM. The
+   * {@link Closeable#close()} method will be called when the program being executed through this program runner
+   * is completed.
+   */
+  private Closeable createClassLoaderCloseable(final Id.Program programId, final RunId runId) {
+    final ClassLoader classLoader = getClass().getClassLoader();
+    return new Closeable() {
+      @Override
+      public void close() throws IOException {
+        Thread t = new Thread("delay-close-thread-" + programId + "-" + runId) {
+          @Override
+          public void run() {
+            Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
```

Comment:
It would be good to add a comment here on why we need a delay before closing the classloader.